include/nbl/asset/CCPUMeshPackerV2.h

```diff
@@ -0,0 +1,222 @@
+
+public:
+    using AllocationParams = MeshPackerBase::AllocationParamsCommon;
+
+    struct AttribAllocParams
+    {
+        size_t offset = INVALID_ADDRESS;
+        size_t size = 0ull;
+    };
+
+	struct ReservedAllocationMeshBuffers
+    {
+        uint32_t mdiAllocationOffset;
+        uint32_t mdiAllocationReservedSize;
+        uint32_t indexAllocationOffset;
+        uint32_t indexAllocationReservedSize;
+        AttribAllocParams attribAllocParams[SVertexInputParams::MAX_VERTEX_ATTRIB_COUNT];
+
+        inline bool isValid()
+        {
+            return this->mdiAllocationOffset != core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
+        }
+    };
+
+    struct PackedMeshBuffer
+    {
+        core::smart_refctd_ptr<video::IGPUBuffer> MDIDataBuffer;
+        core::smart_refctd_ptr<video::IGPUBuffer> vertexBuffer;
+        core::smart_refctd_ptr<video::IGPUBuffer> indexBuffer;
+
+        inline bool isValid()
+        {
+            return this->MDIDataBuffer->getPointer() != nullptr;
+        }
+    };
+
+public:
+    CCPUMeshPackerV2(const AllocationParams& allocParams, uint16_t minTriangleCountPerMDIData = 256u, uint16_t maxTriangleCountPerMDIData = 1024u);
+
+	template <typename Iterator>
+	bool alloc(ReservedAllocationMeshBuffers* rambOut, const Iterator begin, const Iterator end);
+
+    void instantiateDataStorage();
+
+    template <typename Iterator>
+    bool commit(MeshPackerBase::PackedMeshBufferData* pmbdOut, ReservedAllocationMeshBuffers* rambIn, const Iterator begin, const Iterator end);
+
+    PackedMeshBuffer createGPUPackedMeshBuffer(video::IVideoDriver* driver);
+
+public:
+    core::smart_refctd_ptr<ICPUBuffer> m_vtxBuffer;
+    core::smart_refctd_ptr<ICPUBuffer> m_idxBuffer;
+    core::smart_refctd_ptr<ICPUBuffer> m_MDIBuffer;
+
+    const AllocationParams m_allocParams;
+
+};
+
+template <typename MDIStructType>
+CCPUMeshPackerV2<MDIStructType>::CCPUMeshPackerV2(const AllocationParams& allocParams, uint16_t minTriangleCountPerMDIData, uint16_t maxTriangleCountPerMDIData)
+    :IMeshPacker<ICPUMeshBuffer, MDIStructType>(minTriangleCountPerMDIData, maxTriangleCountPerMDIData),
+    m_allocParams(allocParams)
+{
+    initializeCommonAllocators(allocParams);
+}
+
+template <typename MDIStructType>
+template <typename Iterator>
+bool CCPUMeshPackerV2<MDIStructType>::alloc(ReservedAllocationMeshBuffers* rambOut, const Iterator begin, const Iterator end)
+{
+    size_t i = 0ull;
+    for (auto it = begin; it != end; it++)
+    {
+        ReservedAllocationMeshBuffers& ramb = *(rambOut + i);
+        const size_t vtxCnt = (*it)->calcVertexCount();
+        const size_t idxCnt = (*it)->getIndexCount();
+
+        //allocate indices
+        ramb.indexAllocationOffset = m_idxBuffAlctr.alloc_addr(idxCnt, 2u);
+        if (ramb.indexAllocationOffset == INVALID_ADDRESS)
+            return false;
+        ramb.indexAllocationReservedSize = idxCnt * 2;
```

Comment:
A. Use `sizeof(uint16_t)` instead of 2 (side note: never use numerical literals, use stuff that actually has a semantic)
B. You won't get away with this, you need to do the same triangle batching that packer V1 does (because you'll run out of 16bit indices)

Reply:
also dont align to 2, 2 bytes alignment is implied since you're allocating uint16s, not bytes

Reply:
> You won't get away with this, you need to do the same triangle batching that packer V1 does (because you'll run out of 16bit indices)

from what i see, it's done (indices for each batch start from 0), however baseVertex in MDI structs needs to be set accordingly

Reply:
its more about the fact that @Przemog1 reserves the number of indices based on the mesh buffer's index count, but due to splitting meshes he'll get multiple MDI calls per meshbuffer.

Reply:
> 
> 
> also dont align to 2, 2 bytes alignment is implied since you're allocating uint16s, not bytes

an address allocator doesn't know or care about that, it needs to be told alignment explicitly... would work with `alignof(uint16_t)` instead of the `2u`

Reply:
> an address allocator doesn't know or care about that

thats why its implied when im doing 
```
ramb.indexAllocationOffset = alloc_addr();
uint16_t* indexBuffPtr = static_cast<uint16_t*>(m_idxBuffer->getPointer()) + ramb.indexAllocationOffset;
```

Reply:
yeah, that also works..

But @Przemog1 needs to decide which allocators will allocate in "bytes" vs. "structs" (I'd recommend everything except V2's vertex data storage)

Then do through the codebase and make sure the principle is being applied CONSISTENTLY (I see a lot of failures to do this, during my 3 reviews so far)

Reply:
so Przemog get rid of this static_cast<uint16_t*> then...

Reply:
> also dont align to 2, 2 bytes alignment is implied since you're allocating uint16s, not bytes

you right

Reply:
> so Przemog get rid of this static_cast<uint16_t*> then...

wait what? It needs to be there..

Reply:
matt wants you to have single convention for all addr allocators, so that would mean allocating byte-wise and then you have to explicitly specify alignment=2. Dont ask me why

Reply:
see discord